# Introduction to SQL

Here you can access the `books` table used in the course.

--- 
_Note: When using sample integrations such as those that contain course data, you have read-only access. You can run queries, but cannot make any changes such as adding, deleting, or modifying the data (e.g., creating tables, views, etc.)._

## Take Notes

Add notes about the concepts you've learned and SQL cells with queries you want to keep.

_Add your notes here_

In [3]:
-- Add your own queries here
SELECT * 
FROM books

id  ...        genre
0      1.0  ...  Non Fiction
1      2.0  ...      Fiction
2      3.0  ...  Non Fiction
3      4.0  ...      Fiction
4      5.0  ...  Non Fiction
..     ...  ...          ...
347  346.0  ...  Non Fiction
348  347.0  ...      Fiction
349  348.0  ...  Non Fiction
350  349.0  ...      Fiction
351  350.0  ...      Fiction

[352 rows x 5 columns]

Making queries DISTINCT
You've learned that the DISTINCT keyword can be used to return unique values in a field. In this exercise, you'll use this understanding to find out more about the books table!

There are 350 books in the books table, representing all of the books that our local library has available for checkout. But how many different authors are represented in these 350 books? The answer is surely less than 350. For example, J.K. Rowling wrote all seven Harry Potter books, so if our library has all Harry Potter books, seven books will be written by J.K Rowling. There are likely many more repeat authors!

In [4]:
-- Select unique authors from the books table
SELECT DISTINCT author
FROM books;

author
0       John Heilemann
1      Sheryl Sandberg
2         Brené Brown
3                   DK
4       Rebecca Skloot
..                 ...
243        Craig Smith
244    Laurel Randolph
245        Sarah Palin
246  Ann Whitford Paul
247      Joanna Gaines

[248 rows x 1 columns]

In [5]:
-- Select unique authors and genre combinations from the books table
SELECT DISTINCT author, genre
FROM books;

author        genre
0      Eben Alexander  Non Fiction
1       Adam Mansbach      Fiction
2         Garth Stein      Fiction
3       Gillian Flynn      Fiction
4      Tucker Carlson  Non Fiction
..                ...          ...
245       James Comey  Non Fiction
246     Jaycee Dugard  Non Fiction
247      Thug Kitchen  Non Fiction
248  William P. Young      Fiction
249    Deborah Diesen      Fiction

[250 rows x 2 columns]

Aliasing
While the default column names in a SQL result set come from the fields they are created from, you've learned that aliasing can be used to rename these result set columns. This can be helpful for clarifying the intent or contents of the column.

Your task in this exercise is to incorporate an alias into one of the SQL queries that you worked with in the previous exercise!

In [6]:
-- Alias author so that it becomes unique_author
SELECT DISTINCT author AS unique_author
FROM books;

unique_author
0       John Heilemann
1      Sheryl Sandberg
2         Brené Brown
3                   DK
4       Rebecca Skloot
..                 ...
243        Craig Smith
244    Laurel Randolph
245        Sarah Palin
246  Ann Whitford Paul
247      Joanna Gaines

[248 rows x 1 columns]

It's AS easy AS that! Great work. The alias you just implemented makes it clear that only unique authors are listed in the results and that there are no duplicates. This is clear even to someone who is reading only the result set and does not know the SQL code behind the results.

VIEWing your query
You've worked hard to create the below SQL query:

SELECT DISTINCT author AS unique_author
FROM books;
What if you'd like to be able to refer to it later, or allow others to access and use the results? The best way to do this is by creating a view. Recall that a view is a virtual table: it's very similar to a real table, but rather than the data itself being stored, the query code is stored for later use.

In [ ]:
-- Save the results of this query as a view called library_authors
CREATE VIEW library_authors AS
SELECT DISTINCT author AS unique_author
FROM books;

Error: -- Save the results of this query as a view called library_authors
CREATE VIEW library_authors AS
SELECT DISTINCT author AS unique_author
FROM books; - permission denied for schema public

In [ ]:
-- Select all columns from library_authors
SELECT *
FROM library_authors;

In [ ]:
-- Limiting results
-- Let's take a look at a few of the genres represented in our library's books.

-- Recall that limiting results is useful when testing code since result sets can have thousands of results! Queries are often written with a LIMIT of just a few records to test out code before selecting thousands of results from the database.

-- Let's practice with LIMIT!

In [7]:
-- Select the first 10 genres from books using PostgreSQL
SELECT genre
FROM books
LIMIT 10;

genre
0  Non Fiction
1      Fiction
2  Non Fiction
3      Fiction
4  Non Fiction
5      Fiction
6      Fiction
7      Fiction
8  Non Fiction
9      Fiction

## Explore Datasets
Use the `books` table to explore the data and practice your skills!
- Select only the `title` column.
- Alias the `title` column as `book_title`.
- Select the distinct author names from the `author` column.
- Select all records from the table and limit your results to 10.